### PR TITLE
Using isOptional instead of the deprecated hasOptionalKeyword

### PR DESCRIPTION
**Pull request by swaminathanmanish**
**Problem**:
hasOptionalKeyword has been deprecated due to which protobuf null handling was also not working correctly.  
PR that marked it deprecated
https://github.com/protocolbuffers/protobuf/commit/d6157f7c7ec78a6942cfd6d96d2e0b40e007ae0d
We use isOptional() instead that checks only the field LABEL

` /** Is this field declared optional? */
    public boolean isOptional() {
      return proto.getLabel() == FieldDescriptorProto.Label.LABEL_OPTIONAL;
    }`

Reference - https://github.com/protocolbuffers/protobuf/blob/d6157f7c7ec78a6942cfd6d96d2e0b40e007ae0d/java/core/src/main/java/com/google/protobuf/Descriptors.java

**Test:**
Ingested Kafka protobuf stream that has defaults and verified that they are treated as nulls through query (attached image)

Screenshot 2023-09-25 at 2 06 30 PM

**Note:** Proto3 treats the field as optional, when the value is set to its default ([issue](https://github.com/protocolbuffers/protobuf/issues/10560.))

How to use optional field - https://protobuf.dev/programming-guides/proto3/

### CODE REVIEW

File: pinot-plugins/pinot-input-format/pinot-protobuf/src/test/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractorTest.java [hunk 1]
```diff
@@ -166,7 +166,7 @@ private Map<String, Object> createRecord1() {
     record.put(LONG_FIELD, 100L);
     record.put(DOUBLE_FIELD, 1.1);
     record.put(FLOAT_FIELD, 2.2f);
-    record.put(BOOL_FIELD, "false");
+    record.put(BOOL_FIELD, "true");
```
Comment (ege-st) [inline pinot-plugins/pinot-input-format/pinot-protobuf/src/test/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractorTest.java:169]: Why was this changed from `false` to `true`?

**Review by xiangfu0 — APPROVED** [review]

File: pinot-plugins/pinot-input-format/pinot-protobuf/src/test/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractorTest.java (hunk 1 above)
Comment (ege-st) [reply pinot-plugins/pinot-input-format/pinot-protobuf/src/test/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractorTest.java:169]: Taking a look at the test code, if we change this to `true` then none of the tests will verify that we correctly parse a Protobuf message with a boolean field whose value is `false`.

Comment (swaminathanmanish) [reply pinot-plugins/pinot-input-format/pinot-protobuf/src/test/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractorTest.java:169]: Yes I thought about this as well.  Protobuf seems to treat the field as optional, when the value is set to its default. This seems to be the behavior. Whats your suggestion to handle this ?  I found a similar issue in the community. 
https://github.com/protocolbuffers/protobuf/issues/10560

Comment (swaminathanmanish) [reply pinot-plugins/pinot-input-format/pinot-protobuf/src/test/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractorTest.java:169]: > Yes I thought about this as well. Protobuf seems to treat the field as optional, when the value is set to its default. This seems to be the behavior. Whats your suggestion to handle this ? I found a similar issue in the community. [protocolbuffers/protobuf#10560](https://github.com/protocolbuffers/protobuf/issues/10560)

Comment (ege-st) [reply pinot-plugins/pinot-input-format/pinot-protobuf/src/test/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractorTest.java:169]: To make sure I'm not missing anything: if we set a boolean field (which is not optional) to `false` then the PB API will say that it is optional?

Comment (swaminathanmanish) [reply pinot-plugins/pinot-input-format/pinot-protobuf/src/test/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractorTest.java:169]: > To make sure I'm not missing anything: if we set a boolean field (which is not optional) to `false` then the PB API will say that it is optional?

Yes thats correct. Thats what I observed and also read from other sources.

Comment (ege-st) [reply pinot-plugins/pinot-input-format/pinot-protobuf/src/test/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractorTest.java:169]: Ok, then we should have a separate test to confirm the optional that uses a different message, that way we can properly test that we extract `false` from a message.

**Comment by gortiz** [comment]
I'm not an expert in protobuf. Reading about this, it seems to be a common issue to map nulls in proto3. I would say that proto3 designers expected to use `oneof` instead of optional to map null values. That is the protobuf equivalent to `@Nullable T` vs `Optional<T>` in Java.

I assume we cannot decide to use `oneof` because this is something the customers decide, right?

For the context, I think https://github.com/protocolbuffers/protobuf/blob/main/docs/field_presence.md and https://github.com/protocolbuffers/protobuf/blob/main/docs/implementing_proto3_presence.md are very useful to understand the problem.

File: pinot-plugins/pinot-input-format/pinot-protobuf/src/main/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractor.java [hunk 2]
```diff
@@ -54,7 +54,7 @@ public void init(@Nullable Set<String> fields, RecordExtractorConfig recordExtra
   }
 
   private Object getFieldValue(Descriptors.FieldDescriptor fieldDescriptor, Message message) {
-    if (fieldDescriptor.isRepeated() || !fieldDescriptor.hasOptionalKeyword() || message.hasField(fieldDescriptor)) {
+    if (fieldDescriptor.isRepeated() || !fieldDescriptor.isOptional() || message.hasField(fieldDescriptor)) {
```
Comment (gortiz) [inline pinot-plugins/pinot-input-format/pinot-protobuf/src/main/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractor.java:57]: Are strings and bytes marked as isRepeated by default? Reading the proto documentation it seems we cannot distinguish between null and empty value in strings and bytes.

Comment (swaminathanmanish) [reply pinot-plugins/pinot-input-format/pinot-protobuf/src/main/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractor.java:65]: isRepeated is only for collections. We dont want to have the null behavior for collections.

Comment (swaminathanmanish) [reply pinot-plugins/pinot-input-format/pinot-protobuf/src/main/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordExtractor.java:65]: > Are strings and bytes marked as isRepeated by default? Reading the proto documentation it seems we cannot distinguish between null and empty value in strings and bytes.

isRepeated is only for collections. We dont want to have the null behavior for collections.

**Comment by swaminathanmanish** [comment]
> I'm not an expert in protobuf. Reading about this, it seems to be a common issue to map nulls in proto3. I would say that proto3 designers expected to use `oneof` instead of optional to map null values. That is the protobuf equivalent to `@Nullable T` vs `Optional<T>` in Java.
> 
> I assume we cannot decide to use `oneof` because this is something the customers decide, right?
> 
> For the context, I think https://github.com/protocolbuffers/protobuf/blob/main/docs/field_presence.md and https://github.com/protocolbuffers/protobuf/blob/main/docs/implementing_proto3_presence.md are very useful to understand the problem.

We discussed offline. The conclusion is that we go with the proto3 behavior, where addition of 'optional' keyword, helps do the presence check and we can keep the explicitly assigned default. Otherwise a field that has default value explicitly assigned, will fail the presence check and we will treat as null.

https://protobuf.dev/programming-guides/proto3/

**Comment by gortiz** [comment]
Reminder: We need to add tests that make it clear that we are able to read a null when the field has bean cleaned and the default value when it was set to the default value.

**Comment by ege-st** [comment]
> Reminder: We need to add tests that make it clear that we are able to read a null when the field has bean cleaned and the default value when it was set to the default value.

Yes, those tests need to be added asap or we will never get them added.
